release.sh

```diff
@@ -1,16 +1,95 @@
 #!/bin/bash
 
+# Builds and uploads a release, optionally creating a tag for that release.
+# --
+
+syntax() {
+    echo "Syntax: ./release.sh -r <tagname or 'snapshot'> [-f]"
+    echo "Arguments:"
+    echo "  -r: The version string of the SDK to be used in gradle and in the tag name, or '-r snapshot' to publish a snapshot of master."
+    echo "  -f: Force the tag creation if the tag already exists. (not applicable to '-r snapshot')"
+}
+
 # Exit immediately on failure:
 set -e
 
 REPO_ROOT_DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
 cd $REPO_ROOT_DIR
 
+GRADLE_BUILD_FILE=$REPO_ROOT_DIR/build.gradle
+
+FLAG_RELEASE=""
+FLAG_FORCE="false"
+while getopts "r:f" opt; do
+    case $opt in
+        r)
+            FLAG_RELEASE=${OPTARG,,} # to lowercase
+            ;;
+        f)
+            FLAG_FORCE="true"
+            ;;
+        \?)
+            syntax
+            exit 1
+            ;;
+    esac
+done
+if [ -z "$FLAG_RELEASE" ]; then
+    echo "Missing required tag argument (-r <tagname or 'snapshot'>)"
+    syntax
+    exit 1
+fi
+echo "Settings: tag=$FLAG_RELEASE force=$FLAG_FORCE"
+
+export SDK_VERSION=$(egrep "version = \"(.*)\"" $GRADLE_BUILD_FILE | cut -d '"' -f2)
+if [ -z "$SDK_VERSION" ]; then
+    echo "Unable to extract SDK version from $GRADLE_BUILD_FILE"
+    exit 1
+fi
+echo "Current gradle version: $SDK_VERSION"
+
+if [ "x${FLAG_RELEASE}" = "xsnapshot" ]; then
+    # Sanity check: Validate that build.gradle already has a '-SNAPSHOT' version (case sensitive comparison)
+    if [[ "$SDK_VERSION" != *"-SNAPSHOT"* ]]; then
+        echo "SDK version in $GRADLE_BUILD_FILE doesn't contain '-SNAPSHOT'."
+        echo "The master branch must only have snapshot builds."
+        exit 1
+    fi
+else
+    # Uploading a release: Create and publish the release tag, which contains a commit updating build.gradle.
+    # Sanity check: Tagged releases of snapshots is disallowed (case insensitive comparison)
+    if [[ "${FLAG_RELEASE^^}" = *"SNAPSHOT"* ]]; then
+        echo "Requested tag release version may not contain 'SNAPSHOT'."
+        exit 1
+    fi
+    sed -i "s/version = \\\"${SDK_VERSION}\\\"/version = \\\"${FLAG_RELEASE}\\\"/" $GRADLE_BUILD_FILE
```

Comment:
Agreed. This script is doing a lot of cmdline interaction with all the `git` calls though